### PR TITLE
Reorganize CI and add per-crate CI checks

### PR DESCRIPTION
**Pull request by jdisanti**
## Motivation and Context

**Reorganizing CI:** Found a way to reduce copy/pasted setup config in the CI configs by using matrices. Also took this as an opportunity to rename all the workflows and walk away from the tiered terminology that we no longer use.

**Per-crate checks:** This adds the ability to run one-off checks against any runtime crate during the Rust Runtime CI. This can be used for checking the dependency graph for duplicate or undesired dependencies, or testing certain combinations of crate features.

Merging this will require updating the required CI Actions in the branch protection rules.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/ci-sdk.yaml [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        test:
+        - name: Unit Tests
+          run: cargo test $(cat service-with-tests)
+          working-directory: aws-sdk
+        - name: Docs
+          run: cargo doc --no-deps --document-private-items
+          working-directory: aws-sdk
+        - name: Clippy
+          run: cargo clippy
+          working-directory: aws-sdk
```
Comment (rcoh) [inline .github/workflows/ci-sdk.yaml:70]: this is super clever

File: tools/additional-per-crate-checks.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+#
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0.
+#
+
+if [[ $# -ne 1 ]]; then
```
Comment (rcoh) [inline tools/additional-per-crate-checks.sh:7]: I must be missing something, but I don't see this actually being run in CI?

Comment (jdisanti) [reply tools/additional-per-crate-checks.sh:7]: It's at the very end of ci.yaml, as part of runtime tests.

Comment (rcoh) [reply tools/additional-per-crate-checks.sh:7]: Oh! Github didn't expand that file and I thought it was deleted. Didn't realize it was edited 🤕

**Review by rcoh — APPROVED** [review]

**Pull request merged**